src/lib.rs

```diff
@@ -3691,30 +3675,74 @@ impl Build {
         Ok(ret)
     }
 
-    fn apple_deployment_version(&self, os: AppleOs, arch_str: Option<&str>, sdk: &str) -> String {
+    fn apple_deployment_version(&self, os: AppleOs, sdk_parts: &AppleSdkTargetParts) -> String {
         let default_deployment_from_sdk = || {
+            let is_catalyst = matches!(sdk_parts.arch, AppleArchSpec::Catalyst(_));
+            // Needs to be both to distinguish between two targets inside the same SDK, like catalyst in the mac SDK.
+            let cache_key = (os, sdk_parts.arch);
+
             let mut cache = self
                 .apple_versions_cache
                 .lock()
                 .expect("apple_versions_cache lock failed");
 
-            if let Some(ret) = cache.get(sdk) {
+            if let Some(ret) = cache.get(&cache_key) {
                 return Some(ret.clone());
             }
 
-            let version = run_output(
-                self.cmd("xcrun")
-                    .arg("--show-sdk-version")
-                    .arg("--sdk")
-                    .arg(sdk),
-                "xcrun",
-                &self.cargo_output,
-            )
-            .ok()?;
+            let sdk_root = self
+                .apple_sdk_root(&sdk_parts.sdk)
+                .map(PathBuf::from)
+                .map_err(|e| {
+                    self.cargo_output
+                        .print_warning(&format_args!("{}, SDK default version unusable", e));
+                    e
+                })
+                .ok()?;
 
-            let version = std::str::from_utf8(&version).ok()?.trim().to_owned();
+            let sdk_info = fs::read_to_string(sdk_root.join("SDKSettings.json"))
+                .map_err(|e| {
+                    self.cargo_output.print_warning(&format_args!(
+                        "Failed to read SDK properties ({}), SDK default version unusable",
+                        e
+                    ));
+                    e
+                })
+                .ok()?;
+
+            let mut sdk_parser = json::Reader::new(&sdk_info);
+
+            let version = sdk_parser
+                .read_str_from_object("DefaultDeploymentTarget", None)
+                .ok()
+                .or_else(|| {
+                    self.cargo_output.print_warning(&format_args!(
```

Comment:
There're quite a few map_err that makes them hard to read.

Can you extract them into a function that returns Result, then match, log and convert it to Option there?